### PR TITLE
Expose endpointOverride through ServiceClientConfiguration

### PR DESCRIPTION
**Pull request by davidh44**
## Motivation and Context

Expose endpointOverride of SdkClient, through ServiceClientConfiguration

## Modifications

Added endpointOverride to ServiceClientConfiguration

## Testing

added unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/AsyncClientBuilderClass.java [hunk 1]
```diff
@@ -120,13 +121,21 @@ private MethodSpec endpointProviderMethod() {
...
         return MethodSpec.methodBuilder("buildClient")
                          .addAnnotation(Override.class)
-                         .addModifiers(Modifier.PROTECTED, Modifier.FINAL)
+                         .addModifiers(Modifier.PROTECTED)
                          .returns(clientInterfaceName)
                          .addStatement("$T clientConfiguration = super.asyncClientConfiguration()", SdkClientConfiguration.class)
                          .addStatement("this.validateClientOptions(clientConfiguration)")
+                         .addStatement("$T endpointOverride = null", URI.class)
+                         .addCode("if (clientConfiguration.option($T.ENDPOINT_OVERRIDDEN) != null"
+                                  + "&& clientConfiguration.option($T.ENDPOINT_OVERRIDDEN)) {"
```
Comment (millems) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/AsyncClientBuilderClass.java:130]: Nit: clientConfiguration.option($T.ENDPOINT_OVERRIDDEN) == Boolean.TRUE?

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/builder/test-async-client-builder-class.java [hunk 2]
```diff
@@ -28,12 +29,17 @@ public DefaultJsonAsyncClientBuilder tokenProvider(SdkTokenProvider tokenProvide
     }
 
     @Override
-    protected final JsonAsyncClient buildClient() {
+    protected JsonAsyncClient buildClient() {
```
Comment (millems) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/builder/test-async-client-builder-class.java:32]: Why remove final?

File: core/aws-core/src/main/java/software/amazon/awssdk/awscore/AwsServiceClientConfiguration.java [hunk 3]
```diff
@@ -90,19 +92,6 @@ protected BuilderImpl(AwsServiceClientConfiguration awsServiceClientConfiguratio
             this.region = awsServiceClientConfiguration.region();
         }
 
-
-        @Override
-        public Builder overrideConfiguration(ClientOverrideConfiguration clientOverrideConfiguration) {
-            this.overrideConfiguration = clientOverrideConfiguration;
-            return this;
-        }
```
Comment (millems) [inline core/aws-core/src/main/java/software/amazon/awssdk/awscore/AwsServiceClientConfiguration.java:98]: Shouldn't we have this on the interface?

File: core/aws-core/src/main/java/software/amazon/awssdk/awscore/AwsServiceClientConfiguration.java [hunk 4]
```diff
@@ -112,6 +101,11 @@ public final ClientOverrideConfiguration overrideConfiguration() {
         public final Region region() {
             return region;
         }
+
+        @Override
+        public final URI endpointOverride() {
```
Comment (millems) [inline core/aws-core/src/main/java/software/amazon/awssdk/awscore/AwsServiceClientConfiguration.java:106]: Shouldn't we have a setter on the interface to override the return type?

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/AsyncClientBuilderClass.java (hunk 1 above)
Comment (davidh44) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/AsyncClientBuilderClass.java]: updated

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/builder/test-async-client-builder-class.java (hunk 2 above)
Comment (davidh44) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/builder/test-async-client-builder-class.java]: intellij suggested to remove it, added it back

File: core/aws-core/src/main/java/software/amazon/awssdk/awscore/AwsServiceClientConfiguration.java (hunk 4 above)
Comment (davidh44) [reply core/aws-core/src/main/java/software/amazon/awssdk/awscore/AwsServiceClientConfiguration.java:112]: added overridden setters for endpointOverride and overrideConfiguration

File: core/aws-core/src/main/java/software/amazon/awssdk/awscore/AwsServiceClientConfiguration.java (hunk 3 above)
Comment (davidh44) [reply core/aws-core/src/main/java/software/amazon/awssdk/awscore/AwsServiceClientConfiguration.java:98]: added in Builder interface

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/AsyncClientBuilderClass.java [hunk 5]
```diff
@@ -121,16 +121,16 @@ private MethodSpec endpointProviderMethod() {
...
                          .addAnnotation(Override.class)
-                         .addModifiers(Modifier.PROTECTED)
+                         .addModifiers(Modifier.PROTECTED, Modifier.FINAL)
                          .returns(clientInterfaceName)
                          .addStatement("$T clientConfiguration = super.asyncClientConfiguration()", SdkClientConfiguration.class)
                          .addStatement("this.validateClientOptions(clientConfiguration)")
                          .addStatement("$T endpointOverride = null", URI.class)
                          .addCode("if (clientConfiguration.option($T.ENDPOINT_OVERRIDDEN) != null"
-                                  + "&& clientConfiguration.option($T.ENDPOINT_OVERRIDDEN)) {"
+                                  + "&& clientConfiguration.option($T.ENDPOINT_OVERRIDDEN) == $T.TRUE) {"
```
Comment (millems) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/AsyncClientBuilderClass.java:130]: Sorry, I updated my example. `Boolean.TRUE.equals(clientConfiguration.option($T.ENDPOINT_OVERRIDDEN))` is safer/more correct. Can we use that instead?

Comment (davidh44) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/AsyncClientBuilderClass.java]: +1

**Review by millems — APPROVED** [review]

**Pull request merged**